### DISCUSSION TITLE
check_nml_error in fms_mod:  Unknown namelist, or mistyped namelist variable in namelist generic_tracer_nml,   (IOSTAT =           19 )

### DISCUSSION THREAD

**Discussion by amoebaliz** [NOAA-CEFI-Regional-Ocean-Modeling/.github#3] [Common Runtime Errors] [answered]
**Description:** FATAL from PE runtime error. 

**Additional Information:** 
  Current generic tracer namelist:
  
        <namelist name="generic_tracer_nml">
                do_generic_tracer=.true.
                do_generic_abiotic=.false.
                do_generic_CFC=.false.
                do_generic_COBALT=.true.
                do_generic_age=.false.
                do_generic_SF6=.false.
                force_update_fluxes=.true.
                as_param='unused'
                do_vertfill_post=.true.
        </namelist>

**Current Approach:** Check what fms expects from generic tracer namelist to track down discrepancy

**Comment by yichengt900** [comment, marked as answer]
@amoebaliz . Thanks for the info. I have created a feature branch including the new obc fix. If you are following @andrew-c-ross 's xml, could you try the following changes in `mom6_symmetric` component:

```
    <component name="mom6_symmetric" requires="fms" paths="mom6/src/MOM6/config_src/memory/dynamic_symmetric mom6/src/MOM6/config_src/drivers/FMS_cap mom6/src/MOM6/src/*/ mom6/src/MOM6/src/*/*/ mom6/src/MOM6/config_src/external/ODA_hooks mom6/src/MOM6/config_src/external/stochastic_physics mom6/src/MOM6/config_src/external/drifters mom6/src/MOM6/config_src/external/database_comms ocean_BGC/generic_tracers ocean_BGC/mocsy/src mom6/src/MOM6/pkg/GSW-Fortran/*/ mom6/src/MOM6/config_src/infra/FMS1">
      <source versionControl="git" root="https://github.com/NOAA-CEFI-Regional-Ocean-Modeling">
        <codeBase version="$(OCEAN_BGC_GIT_TAG)"> ocean_BGC.git </codeBase>
        <csh><![CDATA[
          git clone -b dev/gfdl https://github.com/NOAA-GFDL/MOM6-examples.git mom6
          pushd mom6
          git checkout $(MOM6_GIT_TAG)
          git submodule update --recursive --init src/MOM6 src/SIS2 src/icebergs
          if ("$(MOM6_GIT_FIX)" != "") then
            (cd src/MOM6 && git remote add ynt https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/MOM6.git && git fetch ynt && git checkout ynt/$(MOM6_GIT_FIX))
          endif
          if ("$(SIS2_GIT_FIX)" != "") then
            (cd src/SIS2 && git checkout $(SIS2_GIT_FIX))
          endif
          popd
        ]]></csh>
      </source>
      <compile>
        <cppDefs><![CDATA[ $(F2003_FLAGS) -DMAX_FIELDS_=100 -DNOT_SET_AFFINITY -D_USE_MOM6_DIAG -D_USE_GENERIC_TRACER  -DUSE_PRECISION=2 -D_FILE_VERSION="'"`git-version-string $<`"'" ]]></cppDefs>
        <!--makeOverrides>OPENMP=""</makeOverrides>  openmp switch for MOM6 may cause crash -->
      </compile>
    </component>
```
With the follwoing tag:

```
  <!-- git tag for general component source code-->
  <property name="RELEASE" value="2022.04"/>

  <!-- git tag for FMS source code -->
  <property name="FMS_GIT_TAG" value="$(RELEASE)"/>

  <!-- git tag for shared source code -->
  <property name="CPL_GIT_TAG" value="$(RELEASE)"/>

  <!-- git tag for shared source code -->
  <property name="NULL_TAG" value="2021.03"/>

  <!-- git tag for ocean_shared source code -->
  <property name="OCEAN_BGC_GIT_TAG" value="dev/cefi"/>

  <!-- git tag date for MOM6 source code -->
  <property name="MOM6_DATE" value="20230201"/>

  <!-- git tag for MOM6 source code -->
  <property name="MOM6_GIT_TAG" value="dev/gfdl"/>

  <!-- git tag for MOM6 source code -->
  <property name="MOM6_GIT_FIX" value="bug/obc_fix"/> <!-- Trying new branch with bug fix for bgc obc index -->

  <!-- SIS2 commit to check out -->
  <property name="SIS2_GIT_FIX" value="0006cd2db5ae1b8cd936d5def129b67c67c565a7"/>

```


I will also conduct test to see how things are going.

**Comment by yichengt900** [comment]
@amoebaliz , I have updated an example xml [here](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/regional-mom6-workflow/blob/bug/obc_fix/NWA12/CEFI_NWA12_cobalt_4P.xml). I tried a quick test run using the NWA12 domain with `SPLIT_RK2B = True` and it seems to work (path of my test case is [here](/lustre/f2/scratch/Yi-cheng.Teng/work/fre/obc_fix/NWA/test) on `gaea`) . Keep in mind you will have to have runtime parameter `SPLIT_RK2B = True` as well as `imbalance_tolerance=1.0e-9` in your generic_COBALT_nml. 

CC @NOAA-CEFI-Regional-Ocean-Modeling/gfdl-med group if you are interested in testing the new obc correction.

Update:

Unfortunately I encounter fatal error: `FATAL from PE     0: write_energy : Ocean velocity has been truncated too many times`  after the first time step integration. I think this is related to the OBC configuration, not just a purely technical issue.

**Reply by amoebaliz** [reply]
Thank you! I am testing it out now. 

I don't appear to read access to your scratch directory on gaea so I'm unable to look at the test case but I've found that Ocean velocity truncation error occurs when REMAP_AUXILIARY_VARS = True. Do you have specified in the MOM_input?

**Reply by amoebaliz** [reply]
Update:

I am currently getting FATAL from PE   ###: ==>biological source/sink imbalance (generic_COBALT_update_from_source): Carbon

I am thinking this may derive from adding in the initial and time-varying boundary esper files

**Reply by yichengt900** [reply]
Oops that's my bad. I am chmod my scratch folder now. Just checked my `MOM_parameter_doc.all` and found `REMAP_AUXILIARY_VARS = False`.

Regarding the bgc source/sink imbalance, maybe you could try i`mbalance_tolerance=1.0e-8` in your `generic_COBALT_nml` namelist to see if you can pass it.

**Reply by amoebaliz** [reply]
still getting the same imbalance error with the larger tolerance - will see if I can isolate obcs vs initial condition file. 

I guess that's REMAP_AUXILIARY_VARS wasn't the issue for the velocity truncation errors. I was getting the same velocity truncation error this morning for two simulations that had been running smoothly until this morning. Both simulations use Andrew's MOM6 fork without the Orlanski fix. I realized I had made several edits to the MOM_input file for the new simulations that would also affect resubmitted jobs that used the shared MOM_input file. So, I made a new MOM_input file for the new simulations and reverted the old simulations back to old MOM_input.  They are both running again without issue. I did a vimdiff on the doc.all files: 

    MOM_parameter_doc.all differences between the failing (working) simulations:
          REMAP_AUXILIARY_VARS = True(False)
          EQN_OF_STATE = "WRIGHT_FULL"("WRIGHT"; includes additional parameter defaults)
          TFREEZE_FORM = "MILLERO_78" ("LINEAR"; includes additional parameter defaults)
    
    SIS_parameter_doc.all differences between the failing (working) simulations:
         DO_RIDGING = True(False)

However, I'm not seeing these settings in your MOM_parameter_doc.all so I guess it's a different issue.

**Reply by yichengt900** [reply]
Thx, @amoebaliz. I gave another try from a run starting from Year 1994 with restarts and this one passed the first 2 days. Will try a longer run just see if I can pass a whole year.

Update: Just checked my NWA12 test run from Year 1994 with `SPLIT_RK2B = True` and it passed. The path on Gaea can be found [here](/lustre/f2/scratch/Yi-cheng.Teng/work/fre/CEFI/NWA/test)
